### PR TITLE
📖 Add a cluster api testing proposal

### PR DESCRIPTION
**Pull request by chuckha**
e2e tests became one of the most requested items from cluster-api users/developers.
This proposal aims to start addressing the problem and figuring out how to solve it.

Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**:
This PR gets started on the testing proposal

I'd like to ping several people who had thoughts about conformance testing. This is not conformance testing yet. We talked about some extremely complex set ups that we are explicitly not targeting for this proposal

Anyone who was interested in this section of the f2f would probably be interested in this proposal

https://docs.google.com/document/d/1t12ItkdlsQC9JvCYnEkU_j7SWC36jTfeWM0H_ZoOd_8/edit#

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: docs/proposals/20190926-cluster-api-testing-framework.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Summary
+
+Cluster API can improve stability and the stability of providers by haivng a pluggable test suite. Providers would be
```
Comment (vincepri) [inline docs/proposals/20190926-cluster-api-testing-framework.md]:
```suggestion
Cluster API can improve stability and the stability of providers by having a pluggable test suite. Providers would be
```

File: docs/proposals/20190926-cluster-api-testing-framework.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
+---
+title: Cluster API testing framework
+authors:
+  - "@chuckha"
+reviewers: []
```
Comment (vincepri) [inline docs/proposals/20190926-cluster-api-testing-framework.md]: Fill this pls

File: docs/proposals/20190926-cluster-api-testing-framework.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
```
Comment (vincepri) [inline docs/proposals/20190926-cluster-api-testing-framework.md]: I'd like to see this filled out to make sure we align on the high level design

File: docs/proposals/20190926-cluster-api-testing-framework.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+* Integrate the framework into Cluster API Provider Docker
+* Integrate the framework into Cluster API Bootstrap Provider Kubeadm
+* Help interested parties adopt the framework
+
+### Non-Goals
+
+* Replace any testing that is specific to a provider
+* To define conformance using these tests
+
+## Proposal
```
Comment (vincepri) [inline docs/proposals/20190926-cluster-api-testing-framework.md]: We should probably add the proposal part as well as the risks of the proposed implementation

**Comment by chuckha** [comment]
@vincepri huh. that feedback is surprising considering the template calls out just the summary and motivation and encourages iteration over a big PR. Do you think those action items can be added as we move forward assuming people agree on the summary and motivation/goals/non-goals or should we update the CAEP template?

**Comment by vincepri** [comment]
@chuckha Not sure, we never merged a proposal without those sections filled out. Usually we suggest for starting up a proposal to be kept in Google Docs for faster iteration between stakeholders and go through a complete review with Cluster API Maintainers once ready.

/cc @ncdc @detiber

**Comment by chuckha** [comment]
ok, it sounds like this is a case of what actually happens vs what is documented to happen is out of date. Can you please update the template's getting started guide to reflect the process you have outlined/were expecting?

**Comment by vincepri** [comment]
I'm interested to see what others think, I don't have a strong opinion and I'm ok for this to be the case going forward

**Comment by neolit123** [comment]
for kubeadm KEPs we do require the design/risks&mitigation sections always filled in before the first merge. implementation-details is nice too. the history section should also have an entry for the first PR merge for the CAEP.

if there are a lot of questions marks for this proposal, then it needs more discussion during meetings.

**Comment by vincepri** [comment]
That sounds reasonable to me (and what we did so far). If others are ok with this approach as well, I'll update the template to make it specific

**Comment by neolit123** [comment]
i wasn't at the discussion or haven't seen a VOD of it, but here is my quick feedback.

if CAPI wants to create a framework for testing CAPI based clusters, i would assume that this framework should be a superset of the k/k test framework.

a repository (CAPI?) can include the framework and build a custom test-suite, similar to the e2e_kubeadm and e2e_node suites that reside in k/k's /test.

a problem with that is that the repository that hosts this framework and suite will now depend on k/k 's /pkg package. there is ongoing work to make the k/k framework not depend on that by the testing-commons group, but this effort is going to take time.

on the other hand if CAPI wants to create a new test framework from scratch, "re-inventing the wheel" is going to be an understatement.

**Comment by detiber** [comment]
> on the other hand if CAPI wants to create a new test framework from scratch, "re-inventing the wheel" is going to be an understatement.

Another approach might be to completely avoid the k/k framework and build the conformance tests as a [sonobuoy plugin](https://sonobuoy.io/docs/v0.16.0/plugins/) instead

**Comment by neolit123** [comment]
> Another approach might be to completely avoid the k/k framework and build the conformance tests as a sonobuoy plugin instead

i think this would mean that we cannot claim Kubernetes conformance for these tests are they will not be part of the official k8s conformance suite. were there discussion of creating Cluster API conformance - e.g. this provider is CAPI conformant?

cc @johnSchnake WDYT about this topic?

**Comment by ncdc** [comment]
@neolit123 "Kubernetes conformance" and "Cluster API conformance (for providers)" are two different topics. CAPI's test framework does not need to intersect with or be the same as k/k's test framework.

**Comment by neolit123** [comment]
@ncdc thanks, that was my question.

would a sonobuoy plugin still need to re-invent a lot of utilities that already exists in the k/k framework?

**Comment by johnSchnake** [comment]
> would a sonobuoy plugin still need to re-invent a lot of utilities that already exists in the k/k framework?

I would assume that if you want to leverage lots of the utilities you'd just vendor all the necessary k/k bits. It is a lot but IMO it unblocks you to just move straight into actually building out a test suite.

You could even then effectively copy the scripts to build/run the tests from the conformance image too. I could help clarify the tiny bit that lets it plug into Sonobuoy if needed (you just need some small yaml which specifies what to run/how as the plugin).

**Comment by neolit123** [comment]
> I would assume that if you want to leverage lots of the utilities you'd just vendor all the necessary k/k bits. It is a lot but IMO it unblocks you to just move straight into actually building out a test suite.

i definitely agree and over time the framework will be more "standalone".

**Comment by neolit123** [comment]
is the plan to have this suite / framework as part of this repository or a separate one?

**Comment by chuckha** [comment]
I'm filling this out and will reopen a PR when we have consensus https://docs.google.com/document/d/1SZkKoDok62eWHg5UVOmc96gW8KET69uNHZ4nWV0yyBs/edit

**Pull request closed without merging**